weasyl/controllers/two_factor_auth.py

```diff
@@ -0,0 +1,251 @@
+
+
+@login_required
+@token_checked
+@twofactorauth_disabled_required
+def tfa_init_verify_post_(request):
+    # Extract parameters from the form
+    verify_checkbox = 'verify' in request.params
+    tfasecret = _get_totp_code_from_session()
+    tfaresponse = request.params['tfaresponse']
+    tfarecoverycodes = _get_recovery_codes_from_session()
+
+    # Does the user want to proceed with enabling 2FA?
+    if verify_checkbox and tfa.store_recovery_codes(request.userid, tfarecoverycodes):
+        # Strip any spaces from the TOTP code (some authenticators display the digits like '123 456')
+        tfaresponse = request.params['tfaresponse'].replace(' ', '')
+
+        # TOTP+2FA Secret validates (activate & redirect to status page)
+        if tfa.activate(request.userid, tfasecret, tfaresponse):
+            # Invalidate all other login sessions
+            invalidate_other_sessions(request.userid)
+            # Clean up the stored session variables
+            _cleanup_session()
+            raise HTTPSeeOther(location="/control/2fa/status")
+        # TOTP+2FA Secret did not validate
+        else:
+            return Response(define.webpage(request.userid, "control/2fa/init_verify.html",
+                            [tfarecoverycodes.split(','), "2fa"]))
+    # The user didn't check the verification checkbox (despite HTML5's client-side check); regenerate codes & redisplay
+    elif not verify_checkbox:
+        return Response(define.webpage(request.userid, "control/2fa/init_verify.html",
+                        [tfarecoverycodes.split(','), "verify"]))
+
+
+@login_required
+@twofactorauth_enabled_required
+def tfa_disable_get_(request):
+    return Response(define.webpage(request.userid, "control/2fa/disable.html",
+                    [define.get_display_name(request.userid), None]))
+
+
+@login_required
+@token_checked
+@twofactorauth_enabled_required
+def tfa_disable_post_(request):
+    tfaresponse = request.params['tfaresponse']
+    verify_checkbox = 'verify' in request.params
+
+    if verify_checkbox:
+        # If 2FA was successfully deactivated... return to 2FA dashboard
+        if tfa.deactivate(request.userid, tfaresponse):
+            raise HTTPSeeOther(location="/control/2fa/status")
+        else:
+            return Response(define.webpage(request.userid, "control/2fa/disable.html",
+                            [define.get_display_name(request.userid), "2fa"]))
+    # The user didn't check the verification checkbox (despite HTML5's client-side check)
+    elif not verify_checkbox:
+        return Response(define.webpage(request.userid, "control/2fa/disable.html",
+                        [define.get_display_name(request.userid), "verify"]))
+
+
+@login_required
+@twofactorauth_enabled_required
+def tfa_generate_recovery_codes_get_(request):
+    recovery_codes = tfa.generate_recovery_codes()
+    _set_recovery_codes_on_session(','.join(recovery_codes))
```

Comment:
Run me through how this situation works with session management?

1. A user opens a browser window and navigates to the screen to generate new recovery codes. They see a list of codes, call it A.
2. They open a second browser window and navigate to the screen to generate new recovery codes there as well. They see a different set of codes. Call it B.
3. They go back to the first browser window and verify that they want to save the recovery codes. They see A on the screen. Which set of codes get saved?

My reading of it is that at step 3 we save a new set of codes inside the user's session. There's one session per cookie so if these actions are done on the same browser, the set of codes that will be saved are B even though it's verified on a screen where A is shown.

On the other hand, if the user uses separate browsers for this (e.g. FireFox and Chromium) with different cookies, then A will be saved.

Although I don't think this will come up too much in practice, it is a little worrisome. The way I could see it being an issue is if a user manages to reload the page at the same time they're issuing their verification post.